### ISSUE TITLE
Use JSON-RPC 2.0-standard notifications in the VM Service

### ISSUE THREAD

**Issue by nex3** [dart-lang/sdk#23897]
The [VM service documentation](https://github.com/dart-lang/sdk/blob/master/runtime/vm/service/service.md) says:

> The Service Protocol is based on JSON-RPC 2.0 (http://www.jsonrpc.org/specification). The Service Protocol has been extended to support pushing events to the client, which is apparently outside the scope of the JSON-RPC specification.

This isn't accurate; JSON-RPC 2.0 does support push events, but it's kind of hidden in the spec. The idea is that both the VM Service itself and the code connecting to it are, in JSON-RPC 2.0 terms, both clients and servers. Events sent by the VM service are formatted like normal JSON-RPC 2.0 method calls, except that they don't have an "id" member and thus don't expect a response.

Changing to a more standard model would be very handy for clients of the VM service. It would allow the standard [`json_rpc_2` package](https://pub.dartlang.org/packages/json_rpc_2) to be used, rather than requiring all clients to implement the protocol from scratch and risk bugs that have already been sussed out.

In practice, the change to the protocol would be relatively minor. Events would change from this:

```json
{
  "event": {
    "type": "Event",
    "kind": "IsolateExit",
    "isolate": {
      "type": "@Isolate",
      "id": "isolates/33",
      "number": "51048743613",
      "name": "worker-isolate"
    }
  },
  "streamId": "Isolate"
}
```

to something like this:

```json
{
  "json-rpc": "2.0",
  "method": "Isolate",
  "params": {
    "type": "Event",
    "kind": "IsolateExit",
    "isolate": {
      "type": "@Isolate",
      "id": "isolates/33",
      "number": "51048743613",
      "name": "worker-isolate"
    }
  }
}
```

You could even preserve backwards compatibility for a time by having an optional parameter to `streamListen` that selects what format events should be sent as.

**Comment by turnidge** [comment]
Sounds good to me.  Will start investigating this.

**Comment by nex3** [comment]
I actually have an in-progress CL to do this. I'm happy to keep cranking on it myself, or I could hand it off to you if you'd prefer.

**Comment by turnidge** [comment]
If you've started, then it would be great to have you do it.

My one request would be to have the response use a "streamNotify" method and pass the streamId and event as parameters, like this:

```
{
  "json-rpc": "2.0",
  "method": "streamNotify",
  "params": {
    "streamId":	"Isolate",
    "event": {
      "type": "Event",
      "kind": "IsolateExit",
      "isolate": {
        "type": "@Isolate",
        "id": "isolates/33",
        "number": "51048743613",
        "name": "worker-isolate"
      }
    }
  }
}
```

Sound good?

**Comment by nex3** [comment]
> My one request would be to have the response use a "streamNotify" method and pass the streamId and event as parameters

Why's that? Since all events will be stream notifications, it seems like it makes sense for them to use the stream ID as the method and the event as the params. It also makes it easier for the client to dispatch different callbacks for different events.

**Comment by turnidge** [comment]
Aesthetically, I think that the stream names make bad method names.  The
rest of the protocol strives to have method names be verbs.  "Isolate",
"Stdout", etc., don't fit my style.  I am willing to make dispatching a bit
more complicated to preserve this.

I plan on adding arbitrary stream ids in the future, for example,
"FutureComplete/isolate/222/object/22" might be a stream id that lets the
debugger subscribe to completions on a specific future.  This stream id
makes an unpleasant method name.

Additionally, if we notify event streams using the "streamNotify" method,
then that leaves the rest of the method namespace open for other
backchannel communication from the server to the client in future versions
of the protocol.  The stream/event model is pretty flexible, but it's nice
to have the option open.

**Comment by nex3** [comment]
Okay, I'll change it over.

**Comment by nex3** [comment]
This was fixed by 61ea0618ce6e29cf101bc9c062222f04b97f211a.
